### ISSUE TITLE
ThreeLetterProtein alphabet can be removed?

### ISSUE THREAD

**Issue by mdehoon** [biopython/biopython#1681]
`Bio/Alphabet/__init__.py` defines a `ThreeLetterProtein` alphabet consisting of three-letter amino acids:
```python
class ThreeLetterProtein(Alphabet):
    """Three letter protein alphabet."""

    size = 3
    letters = [
        "Ala", "Asx", "Cys", "Asp", "Glu", "Phe", "Gly", "His", "Ile",
        "Lys", "Leu", "Met", "Asn", "Pro", "Gln", "Arg", "Ser", "Thr",
        "Sec", "Val", "Trp", "Xaa", "Tyr", "Glx",
        ]

    def _upper(self):
        raise NotImplementedError("We don't have an uppercase three letter "
                                  "protein alphabet.")

    def _lower(self):
        raise NotImplementedError("We don't have a lowercase three letter "
                                  "protein alphabet.")

```

Currently this is only used in `Bio/SeqUtils/__init__.py` to check if a Seq object has this alphabet:
```
$ grep -r ThreeLetterProtein *
Bio/SeqUtils/__init__.py:        elif isinstance(base_alphabet, Alphabet.ThreeLetterProtein):
Bio/Alphabet/__init__.py:class ThreeLetterProtein(Alphabet):
```
```python
    if isinstance(seq, (Seq, MutableSeq)):
        base_alphabet = Alphabet._get_base_alphabet(seq.alphabet)
...
        elif isinstance(base_alphabet, Alphabet.ThreeLetterProtein):
```
but I don't see how a Seq object can have such an alphabet in a meaningful way:

```python
>>> from Bio.Seq import Seq
>>> from Bio.Alphabet import ThreeLetterProtein
>>> seq = Seq(["Arg"], ThreeLetterProtein())
...
TypeError: The sequence data given to a Seq object should be a string (not another Seq object etc)
```

and if we pass the sequence as a single string:
```python
>>> from Bio.Seq import Seq
>>> from Bio.Alphabet import ThreeLetterProtein
>>> seq = Seq("Arg", ThreeLetterProtein())
```

then indexing doesn't work correctly:
```python
>>> seq[0]
'A'
# our alanine has become an arginine
```

I guess that the `ThreeLetterProtein` is a leftover from the past, and can be removed.

**Comment by peterjc** [comment]
From memory it works with ``MutableSeq`` (array based) but not ``Seq`` (string based and therefore only for single-letter alphabets), but I don't see this being used in the test suite which puzzles me. I thought the GA code might have used this...

CC @MarkusPiotrowski as we had some discussion about three-letter alphabet use in ``pairwise2`` not that long ago.

**Comment by mdehoon** [comment]
@peterjc  It is possible to use a three-letter alphabet with MutableSeq, but it's super fragile:
```python
>>> from Bio.Seq import MutableSeq
>>> from Bio.Alphabet import ThreeLetterProtein
>>> seq = MutableSeq(["Arg"], ThreeLetterProtein())
>>> seq[0]
'Arg'
>>> seq = seq.toseq()
>>> seq
Seq('Arg', ThreeLetterProtein())
>>> seq[0]
'A'
```
turning our arginine into alanine again.

**Comment by peterjc** [comment]
You're making a strong case for either deprecating and removing ``ThreeLetterProtein``, or that it needs some love and attention to fix it - or at least make it less broken.

On the bright side, the recently added multiplication support seems to work fine:

```python
>>> from Bio.Seq import MutableSeq
>>> from Bio.Alphabet import ThreeLetterProtein
>>> seq = MutableSeq(["Arg"], ThreeLetterProtein())
>>> seq
MutableSeq('Arg', ThreeLetterProtein())
>>> seq * 4
MutableSeq('ArgArgArgArg', ThreeLetterProtein())
>>> seq *= 4
>>> seq
MutableSeq('ArgArgArgArg', ThreeLetterProtein())
>>> seq[0]
'Arg'
```

Given our agreed long term goal of making the ``Seq`` object (and by extension the other ``Seq``-like objects) more like the Python string, removing this seems simpler.

**Comment by peterjc** [comment]
One of the first fixes should be to the ``MutableSeq``'s ``__repr__`` method to show using arrays, unless it is a single character alphabet where a string-like representation is OK.

Another fix worth considering would be to reject a three-letter-alphabet in ``Seq``'s ``__init__`` method. That would mean clear failures on things like ``mutable_seq.toseq()`` which as in the examples above silently corrupts the data.

**Comment by mdehoon** [comment]
I think MutableSeq should not accept arrays at all. It is undocumented behavior, it's inconsistent with Seq objects, and it lets you put anything into a MutableSeq object:
```python
>>> from Bio.Seq import MutableSeq
>>> m = MutableSeq(['A', 'T', 'U', 'Arg', 'Lys', None, 4, 5.0, ('A', 'C')])
>>> 
```

**Comment by mdehoon** [comment]
@MarkusPiotrowski  To use lists as input for pairwise2, I guess you don't need the ThreeLetterProtein alphabet?

**Comment by MarkusPiotrowski** [comment]
Regarding the ```ThreeLetterProtein``` alphabet I would be in favour of fixing of the bugs. I mean, a three letter protein alphabet can be useful, e.g. for a three letter translation (a feature that all sequence editor programs contain). And we may also see the time, when a 26 letter alphabet is not longer enough to represent all possible members of the molecule group, think of methylated bases, artificial molecules like PNAs, naturally occuring or artificially modified amino acids, etc. 
Having said this, is a list support in sequence objects something one should think of?
I know, this is against the agreed aim to make sequences more string-like. My personal view is that sequence objects should be nearer to the biological molecule they represent, e.g. shouldn't they have additional attributes like "circular" or "single stranded"? However, I'm not a professional bioinformatician, just a plant biologist who likes programming, so my view is from a different angle.

**Comment by MarkusPiotrowski** [comment]
@mdehoon ```pairwise2``` and ```ThreeLetterProtein``` alphabet: No, you don't need it and you *can't* use it, because the Seq object will represent the sequence as a string. If the seq object would return a list (in case of alphabets where the members are not single characters), you could use them.

**Comment by chris-rands** [comment]
Regarding the inputs accepted for `MutableSeq`, I think it should accept arrays as it does currently, because if it doesn't this will break the more efficient integer multiplication methods Peter and I just implemented.

But perhaps `MutableSeq` could also perform some additional `isinstance` check in `__init__` (like `Seq` does). The current the assumption that input is either a string or array is not robust. As mdehoon says currently `MutableSeq` accepts anything and clearly this makes no sense for some types, like integers:

```
>>> from Bio.Seq import MutableSeq
>>> seq = MutableSeq(1)
>>> seq
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/cegg/rands/.local/lib/python3.6/site-packages/Bio/Seq.py", line 1836, in __repr__
    if len(self) > 60:
  File "/home/cegg/rands/.local/lib/python3.6/site-packages/Bio/Seq.py", line 1933, in __len__
    return len(self.data)
TypeError: object of type 'int' has no len()
>>> 
```

**Comment by mdehoon** [comment]
@MarkusPiotrowski  I am aware that pairwise2 can accept lists as input, but that makes the alignment much slower. For example, with a DNA sequence of 200 nucleotides, I am getting these timings:
Bio.Align.PairwiseAligner (using a string): 4 seconds
Bio.pairwise2 (using a string): 21 seconds
Bio.pairwise2 (using a list): 166 seconds
So I would use a different character for modified bases, and run the alignment using a string as input.
If the time comes that a 26 letter alphabet is not sufficient, then using arrays of integers will be much faster than using lists.

**Comment by mdehoon** [comment]
@chris-rands  It's trivial to fix that without having to accept arrays as input.

**Comment by chris-rands** [comment]
@mdehoon Yes the fix would be trivial for integer multiplication like `return self.__class__(str(self) * other, self.alphabet)` instead of [this](https://github.com/biopython/biopython/blob/master/Bio/Seq.py#L2061). However, this would be less efficient that the current implementation working with arrays directly.  That might be a worthwhile trade-off to simplify things, you and Peter would know better than me of course!

**Comment by mdehoon** [comment]
@chris-rands  You could do
```python
s = self.__class__('', self.alphabet)
s.data = self.data * other
return s
```

**Comment by MarkusPiotrowski** [comment]
@mdehoon I think we have (many?) users who use ```Biopython``` only occasionally. They have their locale sequences/formats, etc. Often they use ```Biopython``` for a single task (e.g. alignments) and don't care to much about implementation details. You can of course tell them to convert their sequences into single letters, but isn't 5mC a clearer code than any one-letter abbreviation (are their any official single-letter abbreviations for methylated bases?)? Looking from a naive user perspective, non-single-letter sequences (and accordingly alphabets) would be great. Conversion into memory/speed optimized formats can appear under the hood.

**Comment by chris-rands** [comment]
@mdehoon Good point, that would seem better than what I suggested if we go down this road, but I think perhaps the broader question still remains about whether or not `MutableSeq` should accept arrays to permit multiple character alphabets for the reasons Markus says.

Also, I don't feel `MutableSeq` should be trying to emulate string, it just seems at odds with supporting in-place methods like `append`, `pop` etc., but I don't know the history of Biopython thought on this.

**Comment by peterjc** [comment]
I don't think they use the alphabet objects at all (currently), but there is also three-letter code usage in PDB and SCOP files, thus this mapping:

https://github.com/biopython/biopython/blob/biopython-171/Bio/Data/SCOPData.py#L11

By opening this issue, Michiel has pin pointed one of the rarely used bits of the Alphabet code which does seem like it would benefit from documenting and testing, or removing. There do seem to be some use cases, and on balance I would like to support this. That means explicitly documenting how to use ``MutableSeq`` with a non-single-letter alphabet, and fixing some of the breakages noted here.

What might make sense is for the ``Seq`` and ``UnknownSeq`` object's ``__init__`` to refuse to work with the three-letter-alphabet. This could be done by checking ``if alphabet.size and alphabet.size != 1:`` (see #1683), rather than subclass checking.  ``MutableSeq`` could only enforce this with string input.

Do we need to consider alphabets whose entries are different lengths? e.g. ``A``, ``C``, ``T``, ``G``, ``5mC``, ... would not be ``alphabet.size == 1`` or ``alphabet.size == 3``

On the topic of alphabet checking (are the entries in the sequence all from the alphabet's declared letters?), enforcing this or something equivalent is essential for any clever under the hood optimisation (e.g. 2-bit unambiguous DNA or RNA, 4-bit ambiguous DNA or RNA - something BioJava has long offered). Personally I am of the view that if an explicit alphabet with a defined set of letters is used, the ``Seq`` object ``__init__`` should enforce this.

(This would be a separate issue, but one of the things I would like to get rid of in the alphabet code is the current approach to gaps and stop codons with the strange ``AlphabetEncoder`` class; there ought to be a more elegant way to express this with modern Python.)

**Comment by mdehoon** [comment]
@peterjc  @MarkusPiotrowski  @chris-rands 
I think the first priority is to define what the purpose of alphabets is. There are at least three interpretations:

1. To define how the sequence data is stored internally (i.e. what kind of objects are in seq.data);
2. To define conceptually what the Seq object contains (e.g. this is a protein, or this is DNA, or this is DNA with or without methylation);
3. To define how a Seq object should be presented to the user (e.g. as a single-letter string, a three-letter string, or something else).

Personally it is not clear to me what the purpose of alphabets in Biopython is, which is why I am in favor of simply removing them altogether; I don't think I have ever needed them in 15+ years of using Biopython.

However, if somebody can define the purpose of alphabets, then we can discuss if they are sufficiently useful, and how they should be implemented.

**Comment by peterjc** [comment]
Michiel is right that without a clear vision for how to use alphabet objects in Biopython, it is hard to make strategic choices here.

If you are only dealing with single-letter-alphabets, then (1) and (3) are currently simplified to just following the Python string conventions. Therefore right now my only practical usage is (2), explicit typing of the sequence. e.g. This is used in ``Bio.SeqIO`` for some of the sequence formats.

I like the idea of validating sequence letters matching the alphabet under (1), but worry about performance hits.

Separately, we might want to support other internal storage in addition to strings and arrays - e.g. 2-bit or 4-bit nucleotide sequences, but I think that could be done with or without the current alphabet objects.

One option (which I suspect Michiel would support?) is too deprecate the current alphabet system, and replace it with a minimal typing system (just: nucleotide, DNA, RNA, protein or unknown). This could be exposed as a new attribute of the sequence objects (e.g. ``.type`` although that is a special word in Python), with a backward compatible approach to the initialisation API, e.g. allow ``my_dna = Seq("ACGT", "DNA")`` or ``my_amino_acids = Seq("MXXG*", "protein")``. I am willing to try this on a branch.

Another more extreme option (which I think Michiel would prefer?) is to deprecate and remove the current alphabet system completely (see also #1674 ). The current sequence type annotation requirements in ``Bio.SeqIO`` would have to be moved to the ``SeqRecord`` annotation dictionary instead (e.g. DNA, genomic DNA, RNA, ...).

However, should we be thinking about non-single-letter alphabets? I can see two clear use cases here, neither of which currently affects me personally. First, modified amino acids as seen in protein structure databases, and second, nucleotide sequences with per-base methylation status. We could perhaps still proceed with simplifying the alphabet system as outlined above, then introduce a new (array based?) sequence class for these?

If we ask the Biopython community how many people are using existing non-single-letter-alphabets, like the ``ThreeLetterProtein`` alphabet, the answer will likely be zero - this corner of the code is not documented in the Tutorial, nor covered in the unit tests, so you'd only find it by exploring the source code.

**Comment by mmokrejs** [comment]
Hi,
  although I read the issues #1674 and this #1681 I am a bit lost, mainly because I am not that much into the guts of biopython. So, below are my comments/impressions.


I would not cry if I could replace 

```
from Bio.Alphabet import generic_dna
def create_SeqObj_from_sequence_and_quals(myid, description, sequence, quals, alphabet=generic_dna):
    """Create a simple SeqRecord object from scratch using sequence a quals.

    Work around placeholder values in biopython, see https://github.com/biopython/biopython/issues/831
    """

    if sequence and myid:
        if description:
            _record = SeqRecord(Seq(sequence, alphabet), id=myid, description=description, name='')
        else:
            _record = SeqRecord(Seq(sequence, alphabet), id=myid, description='', name='')
        if quals:
            _record.letter_annotations["phred_quality"]=quals
        return _record
    else:
        return None
```

with

`record = SeqRecord(sequence, quals, 'DNA')`

but I did not test whether the above old code is still needed with current biopython.

I am surprised biopython does not check already if the sequence complies with the alphabet but I do not have time to play with this. I am fine with keeping autoamatic checks for alphabet-compliance disabled, just expose a function like

`check_for_compliance(seq, alphabet)`

I could imagine use of mixed alphabet-sizes for m7C or the mRNA cap structure ( https://en.wikipedia.org/wiki/Five-prime_cap ) sometimes abbreviated as m7G but more properly as m7GpppG if memory serves me right. DNA primers often contain modified nucleotides and it is awkward one cannout use the sequences without mangling:

>V35R doi:10.1371/journal.pone.0027310.t003 aka 926R 16S V5 primer for V35 region (CCGTCAATTCMTTTRAGT) in http://www.hmpdacc.org/doc/16S_Sequencing_SOP_4.2.2.pdf
CCGTCAATTCMTTTRAGT

Further, biotion is often used to tag nucleotides: https://en.wikipedia.org/wiki/Biotinylation , well there are much more.

https://www.thermofisher.com/cz/en/home/life-science/oligonucleotides-primers-probes-genes/custom-dna-oligos/oligo-configuration-options.html#5prime

https://www.thermofisher.com/cz/en/home/life-science/oligonucleotides-primers-probes-genes/custom-dna-oligos/oligo-configuration-options.html#internal

https://www.thermofisher.com/cz/en/home/life-science/oligonucleotides-primers-probes-genes/custom-dna-oligos/oligo-configuration-options.html#3prime



If the aim is to drop the thre-leteer alphabets alltogether from biopython I would be stroingly against. They are more exact (hence preferred) than single-letter alphabets. Also, time to time one needs a tool to convert proteins from single-letter to three-letter representations or back. Or translate DNA into 3-letter protein code or backtranslate from aminoacids into DNA codons although inexactly.

**Comment by peterjc** [comment]
It seems there is interest in using three-letter alphabets, but sadly currently the support for this in ``MutableSeq`` is limited, e.g.

```python
>>> from Bio.Seq import MutableSeq
>>> from Bio.Alphabet import ThreeLetterProtein 
>>> my_protein = MutableSeq(["Arg", "Xxx", "Leu", "Arg"], ThreeLetterProtein())
>>> my_protein
MutableSeq('ArgXxxLeuArg', ThreeLetterProtein())
>>> len(my_protein)
4
>>> my_protein[0]
'Arg'
>>> my_protein[1]
'Xxx'
>>> my_protein[2]
'Leu'
>>> my_protein[3]
'Arg'
>>> my_protein.count("A")
0
>>> my_protein.count("Arg")
2
```

Also ``.pop()``, ``.append()``, ``.extend()`` seem fine if given reasonable inputs. So far so good.

```python
>>> str(my_protein)
'ArgXxxLeuArg'
```

The ``__str__`` behaviour is perhaps slightly surprising, but seems practical for equality, sorting, etc - and easy to work with _provided_ the alphabet entries are all the same length (here 3).

On to the more clear cut problems, ``__repr__`` does not let you re-create the object - although arguably this is a fault in the ``__init__`` not handling the concatenated string as input:

```python
>>> my_protein
MutableSeq('ArgXxxLeuArg', ThreeLetterProtein())
>>> len(MutableSeq('ArgXxxLeuArg', ThreeLetterProtein()))
12
```

Calling ``.toseq()`` corrupts the data:

```python
>>> my_protein.toseq()
Seq('ArgXxxLeuArg', ThreeLetterProtein())
>>> len(my_protein.toseq())
12
```

The count-overlaps implementation fails to reject requests like this:

```
>>> my_protein.count_overlap("XxxLeuA")
1
```

Methods ``.complement()``, ``.reverse_complement()`` do not consider ``ThreeLetterProtein()`` as protein, and thus fail to throw an alphabet based exception - and fail cryptically instead.

These seem fixable - likewise for a fixed width modified nucleotide alphabet. Variable length alphabets would cause trouble (especially in the string like behaviour).

**Comment by mdehoon** [comment]
Does everybody agree with Peter's suggestion that the purpose of alphabets is (2) "To define conceptually what the Seq object contains (e.g. this is a protein, or this is DNA, or this is DNA with or without methylation)" ?

**Comment by MarkusPiotrowski** [comment]
Not really. This could be represented by a simple attribute like "seq_type' or whatever. I thought the main purpose of the alphabet is to list the valid elements of the sequence.

**Comment by peterjc** [comment]
So me the primary purpose of alphabets from the ``Seq`` object perspective is as Michiel put it (2) "To define conceptually what the Seq object contains (e.g. this is a protein, or this is DNA, or this is DNA with or without methylation)"?

That could indeed mostly be captured by a "seq_type" attribute of some sort, although having started experimenting with this on a branch, it isn't quite as simple as I initially hoped. The main hurdle seems to be use of ``.letters`` (as Markus might have realised).

Markus is right that if anyone is actually using the alphabet itself, the main purpose seems to be the expected letters in the sequence (e.g. building dictionaries, frequency tables, etc), be that the four letters ACGT, or an ambiguous nucleotide alphabet or what ever. In practise I suspect most of those use cases can sort to an if/else and import the relevant string of letters from ``Bio.Data`` if needed.

**Comment by mdehoon** [comment]
@MarkusPiotrowski  You mean that the purpose of the alphabet is to check the contents of a Seq object during instantiation? If so, how do you envision this to work in practice?

**Comment by peterjc** [comment]
You could follow something like this, although in the context of the Seq object initialisation using a no-letter generic alphabet would signal no checking of the contents:
https://github.com/biopython/biopython/blob/biopython-171/Bio/Alphabet/__init__.py#L441

**Comment by mdehoon** [comment]
@MarkusPiotrowski  @peterjc  For large sequences, this will take a very long time (2.5 minutes just for human chromosome 1, compared to 4 seconds to read the sequence without validating against the alphabet). Then validating against the alphabet should be optional. How should this be done (meaning what will the API look like)?

**Comment by peterjc** [comment]
Practically speaking there would need to be a way to control validation at the parser level (e.g. when calling ``Bio.SeqIO.parse(...)`` and friends).

This could be as simple as in the code I linked to earlier - if you want validation, explicitly use an alphabet with letters, otherwise use a generic alphabet?

In some cases, the file parser itself can ensure the letters are appropriate (e.g. binary file formats like SFF or BAM), so in those special cases would be expedient to be able to deliberately skip any ``Seq`` object alphabet validation.

**Comment by chris-rands** [comment]
For validation, [my tests](https://pastebin.com/m3fSvKEN) on hg38 chr1 suggest that using `set.issubset` might be slightly faster than the implementation in `_verify_alphabet`, but it's not a proper benchmark of course

**Comment by mdehoon** [comment]
@peterjc  Validation also needs to be controlled when creating a Seq or MutableSeq object by hand.

**Comment by peterjc** [comment]
In that example it would be reasonable to get a ``TypeError`` for incompatible alphabets.

**Comment by mdehoon** [comment]
I would expect Biopython to be a bit more intelligent there. Now this example would require something like
```python
>>> m3_1 = MutableSeq("".join([IUPACData.protein_letters_3to1[aa] for aa in m3]), IUPAC.IUPACProtein())
>>> m1 + m3_1
MutableSeq('MVRK', IUPACProtein())
```
which is way too complicated.

Not to mention that this will happen:
```python
>>> m3 == m3_1
False
```

**Comment by mdehoon** [comment]
@MarkusPiotrowski  Yesterday I had a chance to meet up with @peterjc , and we discussed the issue of alphabets in Biopython. Because of the complexity of alphabets and their use in Biopython, we felt that it may be a good idea to have a PEP (Python Enhancement Proposal)-like discussion to define the purpose of alphabets and their technical implementation in Biopython. This would mean that somebody who is in favor of having alphabets in Biopython would work out a proposal with all the details to allow developers and users to think through the implications. Since alphabets have been under discussion for more than 10 years, we are thinking to put a time limit to the proposal (e.g., until January 1st, 2020), meaning that if no agreement on the proposal is reached by then, alphabets would be removed from Biopython. This would give people who are in favor of alphabets to make their case, while guaranteeing that a conclusion will be reached (either a well-defined and usable alphabet, or no alphabet) within the next ~1.5 years. Details are up for discussion; I'll post this to the mailing list also to keep everybody informed.

**Comment by fabianegli** [comment]
The study of post translational modifications (PTM) is prevalent and we will see an increase of applications that involve sequences with PTMs. The expansion of alphabets and Biopythons ability to cope with them in a straight forward manner is paramount. There are hundreds of PTMs and coding them in multi-letter codes seems to be problematic for some underlying software Biopython uses for alignments. 

Alphabets have long been a pain in the computer world, but it has been mostly alleviated by the use of UTF8. I propose UTF8 as Biopythons internal alphabet encoding. The character space should be big enough to allow encoding all the known posttranslational modifications.

Did the PEP-like discussion start somewhere? If so it should be linked here.

**Comment by peterjc** [comment]
TL;DR - there is no PEP-like proposal for alphabets as yet. Michael wants to remove the currently alphabet objects completely (but bits of the wider codebase still need somewhere to record DNA / RNA / protein), while I lean towards a drastic simplification to just an enum-like type system.

Technical input from someone wanting to work with modified bases / residues would be welcome. Taking advantage of unicode might work better than 3-letter codes.

Longer recap:

Here @mdehoon proposed writing a PEP-like proposal for the alphabets,

https://mailman.open-bio.org/pipermail/biopython/2018-August/016524.html

> Dear all,
> 
> While sequence objects in Biopython have an associated alphabet, the purpose of alphabets in Biopython is currently not well-defined.
> I can imagine these three interpretations of their purpose:
> To define how the sequence data is stored internally in a Seq object (i.e. what kind of objects are in seq.data);
> To define conceptually what the Seq object contains (e.g. this is a protein, or this is DNA, or this is DNA with or without methylation);
> To define how a Seq object should be presented to the user (e.g. as a single-letter string, a three-letter string, or something else).
> (and there may be others that I have overlooked).
> 
> To justify having alphabets as a part of Biopython, their purpose should be clearly defined.
> Because of the complexity of alphabets and their use in Biopython, we felt that it may be a good idea to have a PEP (Python Enhancement Proposal)-like discussion to define the purpose of alphabets and their technical implementation in Biopython. This would mean that somebody who is in favor of having alphabets in Biopython would work out a proposal with all the details to allow developers and users to think through the implications.
> 
> Here you can find a description of PEPs and what should go in them:
> 
> https://www.python.org/dev/peps/pep-0001/
> 
> Not all of it is applicable to Biopython, but it may serve as a general guideline.
> 
> The Alphabet BEP (Biopython Enhancement Proposal) could be hosted on the Biopython website so that everybody can follow the discussion.
> Since alphabets have been under discussion for more than 10 years, we are thinking to put a time limit to the proposal (e.g., until January 1st, 2020), meaning that if no agreement on the proposal is reached by then, alphabets would be removed from Biopython. This would give people who are in favor of alphabets to make their case, while guaranteeing that a conclusion will be reached (either a well-defined and usable alphabet, or no alphabet) within the next ~1.5 years. 
> 
> Any volunteers? Seq objects and therefore their alphabets are a key feature of Biopython, and working through a BEP can give you the opportunity to help design a major part of Biopython.
> 
> Best,
> -Michiel

The thread continued in October, but there have been no concrete proposals.

https://mailman.open-bio.org/pipermail/biopython/2018-October/016555.html

Michiel advocates starting by removing the alphabets:

https://mailman.open-bio.org/pipermail/biopython/2018-October/016559.html

> Hi Peter,
> To keep things manageable, I would suggest to remove Bio.Alphabet first, before considering to introduce a new system to replace it. 
> Also, after removing Bio.Alphabet, we may have a better idea if and how it should be replaced.
> Best,-Michiel
> 

My most recent reply,

https://mailman.open-bio.org/pipermail/biopython/2018-October/016560.html

> HI Michiel,
> 
> My point was that *something* has to replace Bio.Alphabet
> for some of the existing use cases - including various file
> formats in SeqIO and AlignIO which record DNA, RNA
> vs Protein. e.g.
> 
> https://github.com/biopython/biopython/blob/biopython-172/Bio/AlignIO/NexusIO.py#L129
> https://github.com/biopython/biopython/blob/biopython-172/Bio/SeqIO/InsdcIO.py#L613
> https://github.com/biopython/biopython/blob/biopython-172/Bio/SeqIO/SeqXmlIO.py#L348
> 
> This could be a new convention for where to store this
> in the SeqRecord, or as I am currently exploring, a
> simplified alphabet attribute of the Seq object.
> 
> This effort should uncover any other roadblocks to a
> removal or another replacement alphabet system would
> face - and might help with coming up with a better idea.
> 
> Peter

Removal or simplification to a simple enum type system as on https://github.com/peterjc/biopython/tree/alpha_lite would not in itself help with representing modified bases or residues, where 3-letter codes kind of work in ``MutableSeq`` but are not well supported.

Moving to expressing modified bases or residues using unicode (and in particular UTF8) might work well (it should even allowing emoji in sequences), as this has the enormous positives that each symbol has length one when viewed as a (unicode) string. And we've a long term goal that the ``Seq`` object behave more and more like the Python (unicode) string, and ultimately perhaps subclass it.

This idea could probably be done with the current alphabet objects, or a simplified alphabet type enum, or no alphabet attribute.

**Comment by mdehoon** [comment]
@fabianegli @peterjc 
UTF8 would be slow, because the number of bytes used to store a character varies. Then if `s` is a Seq object,
```python
>>> c = s[1000000]
```
requires evaluating one million characters to find out where s[1000000] starts.

**Comment by peterjc** [comment]
Currently on Python 3 the ``Seq`` object is using a Python unicode string, so we are implicitly using the system default encoding, i.e. it _already_ is usually stored UTF-8.

On the bright side, almost all the biological sequences we use currently are plain ASCII, and Python 3 handles those very efficiently as unicode strings - we ought not to need to worry about performance for now.

If we start dealing with entire chromosomes/genomes with non-ASCII, it might be a problem (e.g. methylation state - but there you could probably still pick an ASCII-IUPAC-DNA + other ASCII character alphabet).

+1 on opening a new issue for storing non-IUPAC characters using non-ASCII characters available.

**Comment by mdehoon** [comment]
@peterjc  As far as I know, Python3 does not use UTF-8 internally.

**Comment by peterjc** [comment]
I may have misunderstood something then, or am being imprecise with my language and the terms (unicode, encoding, UTF-8 etc):
https://docs.python.org/3/howto/unicode.html

**Comment by peterjc** [comment]
Somewhat relevant news from the PyPy project, https://twitter.com/pypyproject/status/1095971192513708032

> unicode-utf8 just got merged! Unicode strings are now internally represented as utf-8 in PyPy, with an optional extra index data structure to make indexing O(1). We'll write a blog post about it eventually.

**Comment by fabianegli** [comment]
Accessing char at pos x in a string is also O(1) in Python 3. Although memory usage for UTF-8 strings is higher than for ASCII. The code point size in Python is fixed - check out wide and narrow builds. This is also the case - to some extent - in the new PyPy, when the additional index is used for the linear time access. Thus I anticipate that, if at all, memory usage is a more serious concern than speed, if it is one at all.

**Comment by fabianegli** [comment]
After following this issue and considering the implications of my proposal to use UTF-8 to extended alphabets and encodings for bio-sequences, I am currently setting up a project to investigate this proposal. I chose to make the process public to speed it up and allow stakeholders to influence the project or collaborate. Here is the link to the (still rudimentary) repository: https://github.com/fabianegli/bioenc

**Comment by peterjc** [comment]
Closing as part of #2046 - we're now in the process of removing all of ``Bio.Alphabet`` now.
